### ISSUE TITLE
single server browser list for LAN and Internet servers

### ISSUE THREAD

**Issue by illwieckz** [Unvanquished/Unvanquished#1112]
Just put LAN servers above Internet servers if they exist.

Like [Quetoo](http://quetoo.org/):

[![quetoo server browser](https://dl.illwieckz.net/b/quetoo/screenshots/20171227-121016-000.quetoo-server-browser.png)](https://dl.illwieckz.net/b/quetoo/screenshots/20171227-121016-000.quetoo-server-browser.png)

Simplifies UI.

This requires changes in [unvanquished_src.dpkdir](http://github.com/UnvanquishedAssets/unvanquished_src.dpkdir) too.

**Comment by Dimitriio** [comment]
Hello @illwieckz,
i am working on it. I have done for now the UI. As you can see, i removed the tabs and have added the "Source" column.
But i have an issue with the data. For now it is split in 2 arrays. And it seems quite difficult merging them without changing how it's done as all events are called through the generic `dataSourceCmd_t`.
![Screenshot from 2020-07-18 22-25-45](https://user-images.githubusercontent.com/9897777/87861419-56d64c80-c946-11ea-950c-0b5a767fb874.png)

**Comment by illwieckz** [comment]
Hi, that's nice to see! :+1: 

Do we need to merge them? We may just list LAN servers atop Internet server list, but in this case the issue is: how to not have two heading cells (“Name, Source, Map…”), and we may also have a scrolling issue: if we do two lists we may want to scroll a kind of frame containing the two lists and not to scroll the lists themselves…

**Comment by Dimitriio** [comment]
Well, the source of the datagrid is only one array, so if we want one datagrid to be shown (with the sorting), we need to merge them, atleast in the exposed data that the grid will use. 

For now the source can be `server_browser.internet` or `server_browser.local` so to show both, it needs some adjustements. The model is inside the variable `rocketInfo.data.servers[ netSrc ]` where netSrc correspond to `AS_LOCAL 0`, `AS_GLOBAL 1`, `AS_FAVORITES 2`

And what could be bothering too is both filtering & sorting if we consider using 2 datagrid.

We could keep `rocketInfo.data.servers[ netSrc ]` and just directly add both LAN & internet to the data but we will loose the capabilities to do it separatly for only internet or local in the future. And it won't have the same comportment as exec, build, clean, ect that will still need the `netSrc` to be working (except if we remove the generecity of all methods, but i dunno the actual impacts).

**Comment by slipher** [comment]
As @DolceTriade tells me, the right way to do this stuff is to have a Lua script that builds the table, and some presentation-agnostic way of making the data to Lua. Unfortunately, there are no existing examples of doing it this way in our codebase.

**Comment by Dimitriio** [comment]
Hi @slipher, @DolceTriade,
I saw in the sample `luainvaders` of librocket something that look like what you were saying.
The example show a Lua Interface exposing C++ method usable in lua script :
```


void LuaInterface::Initialise(lua_State* L)
{
    InitGame(L);
    Rocket::Core::Factory::RegisterElementInstancer("game",new ElementGameInstancer())->RemoveReference();
}

void LuaInterface::InitGame(lua_State *L)
{
    luaL_dostring(L,"Game = Game or {}"); //doing this in Lua because it would be 10+ lines in C++
    lua_getglobal(L,"Game");
    int game = lua_gettop(L);

    if(lua_isnil(L,game))
    {
        Rocket::Core::Log::Message(Rocket::Core::Log::LT_ERROR, "Error creating the Game table from C++ in LuaInterface::InitGame");
        return;
    }
    
    //this can be done in a loop similar to the implementation of LuaType::_regfunctions, but I am
    //defining them explicitly so that there is no black magic
    lua_pushcfunction(L,GameShutdown);
    lua_setfield(L, game, "Shutdown");
}
```
In the .rml template :
```
<rml>
[...]
		<button onclick="import game;game.Shutdown()">Exit</button>
[...]
</rml>
```

Is it what you propose to do ?

**Comment by slipher** [comment]
Partially. This shows registering a C++ callback (albeit in a weird way, I think there are utilities for doing that), and making an event handler that uses the callback. So the ideal is that you'd have a C++ callback that returns the server data or whatever. And the Lua code would build the table using that data. So the C++ code would no longer need to know anything about how the table looks.

Unfortunately the server browser case is extra annoying, because I think the data is fetched asynchronously, meaning you can't get the data right away when someone clicks the update list button. So the Lua code would have to be polling for changes every frame, or the C++ code would need to make a data change event, or something.

**Comment by DolceTriade** [comment]
ya, we should start moving as much as we can into Lua, however, in this case, i don't think we have the infrastructure there yet, so we can just merge the two sources into a single array.
